apps/mark-scan/fai-100-controller/src/fai_100_controllerd.rs

```diff
@@ -0,0 +1,555 @@
+    );
+    let mut keyboard = create_keyboard("FAI 100 Accessible Controller Daemon Virtual Device")?;
+    log!(
+        event_id: EventId::CreateVirtualUinputDeviceComplete,
+        disposition: Disposition::Success,
+        event_type: EventType::SystemAction
+    );
+
+    log!(
+        EventId::ControllerConnectionInit;
+        EventType::SystemAction
+    );
+
+    if let Some(mut port) = get_usb_device() {
+        log!(
+            event_id: EventId::Info,
+            disposition: Disposition::Success,
+            event_type: EventType::SystemStatus,
+            message: ("Connected to port successfully").to_string()
+        );
+
+        if let Err(err) = validate_connection(&mut port) {
+            log!(
+                event_id: EventId::ControllerHandshakeComplete,
+                message: err.to_string(),
+                event_type: EventType::SystemAction,
+                disposition: Disposition::Failure
+            );
+
+            exit(1);
+        }
+
+        run_event_loop(&mut port, &running, &mut keyboard);
+        exit(0);
+    }
+
+    if args.skip_hardware_check {
+        run_no_op_event_loop(&running);
+        exit(0);
+    }
+
+    log!(
+        event_id: EventId::ProcessTerminated,
+        event_type: EventType::SystemStatus,
+        message: "Could not connect to ATI controller".to_string()
+    );
+    exit(1);
+}
+
+fn get_usb_device() -> Option<UsbDevice> {
+    match UsbDevice::open_by_ids(FAI_100_VID, FAI_100_PID) {
+        Ok(port) => {
+            log!(
+                event_id: EventId::ControllerConnectionComplete,
+                event_type: EventType::SystemAction,
+                disposition: Disposition::Success
+            );
+
+            return Some(port);
+        }
+        Err(e) => {
+            log!(
+                event_id: EventId::ControllerConnectionComplete,
+                message: format!("Failed to open controller device. Error: {e}"),
+                event_type: EventType::SystemAction,
+                disposition: Disposition::Failure
+            );
+        }
+    }
+
+    None
+}
+
+fn validate_connection(usb_device: &mut UsbDevice) -> Result<(), io::Error> {
+    log!(
+        event_id: EventId::ControllerHandshakeInit,
+        event_type: EventType::SystemAction
+    );
+
+    let version_cmd = create_command!(GetFirmwareVersion);
+    let version_cmd: Vec<u8> = version_cmd.into();
+
+    match usb_device.write_bulk(&version_cmd) {
+        Ok(num_bytes) => log!(
+            event_id: EventId::Info,
+            message: format!("validate_connection: {num_bytes} bytes written"),
+            event_type: EventType::SystemAction
+        ),
+        Err(e) => log!(
+            event_id: EventId::UnknownError,
+            message: format!("validate_connection unexpected error when writing command: {e:?}"),
+            event_type: EventType::SystemStatus
+        ),
+    }
+
+    let mut buf: [u8; BUFFER_MAX_BYTES] = [0; BUFFER_MAX_BYTES];
```

Comment:
No type annotation needed:
```suggestion
    let mut buf = [0; BUFFER_MAX_BYTES];
```